src/ui/index.ts

```diff
@@ -102,27 +104,79 @@ export function getNodeId({ attributes }: UiNode) {
 }
 
 /**
- * Filters nodes by their groups.
+ * Return the node input attribute type
+ * In <input> elements we have a variety of types, such as text, password, email, etc.
+ * When the attribute is null or the `type` attribute is not present, we assume it has no defined type.
+ * @param attr
+ * @returns type of node
+ */
+export const getNodeInputType = (attr: UiNodeAttributes): string =>
+  attr && "type" in attr ? attr.type : ""
+
+export type FilterNodesByGroups = {
+  nodes: Array<UiNode>
+  groups?: Array<UiNodeGroupEnum | string> | UiNodeGroupEnum | string
+  withoutDefaultGroup?: boolean
+  attributes?:
+    | Array<UiNodeInputAttributesTypeEnum | string>
+    | UiNodeInputAttributesTypeEnum
+    | string
+  withoutDefaultAttributes?: boolean
+  excludeAttributes?:
+    | Array<UiNodeInputAttributesTypeEnum | string>
+    | UiNodeInputAttributesTypeEnum
+    | string
+}
+
+/**
+ * Filters nodes by their groups and attributes.
  *
  * Will always add default nodes unless `withoutDefaultGroup` is true.
- *
+ * Will always add default attributes unless `withoutDefaultAttributes` is true.
  * @param nodes
  * @param groups
  * @param withoutDefaultGroup
+ * @param attributes
+ * @param withoutDefaultAttributes
  */
-export const filterNodesByGroups = (
-  nodes: Array<UiNode>,
-  groups?: Array<string> | string,
-  withoutDefaultGroup?: boolean
-) => {
-  if (!groups || groups.length === 0) {
-    return nodes
-  }
-
-  const search = typeof groups === 'string' ? groups.split(',') : groups
-  if (!withoutDefaultGroup) {
-    search.push('default')
-  }
+export const filterNodesByGroups = ({
+  nodes,
+  groups,
+  withoutDefaultGroup,
+  attributes,
+  withoutDefaultAttributes,
+  excludeAttributes,
+}: FilterNodesByGroups) => {
+  const search = (s: Array<string> | string) =>
+    typeof s === "string" ? s.split(",") : s
 
-  return nodes.filter(({ group }) => search.indexOf(group) > -1)
+  return nodes
+    .filter(({ group }) => {
+      if (!groups) return true
+      const g = search(groups)
+      if (!withoutDefaultGroup) {
+        g.push("default")
+      }
+      return g.indexOf(group) > -1
+    })
+    .filter(({ group, attributes: attr }) => {
+      if (!attributes) return true
+      const a = search(attributes)
+      if (!withoutDefaultAttributes) {
+        // always add hidden fields e.g. csrf
+        if (group.includes("default")) {
+          a.push("hidden")
+        }
+        // automatically add the necessary fields for webauthn and totp
+        if (group.includes("webauthn") || group.includes("totp")) {
+          a.push("input", "script")
+        }
+      }
+      return a.indexOf(getNodeInputType(attr)) > -1
+    })
+    .filter(({ attributes: attr }) => {
+      if (!excludeAttributes) return true
+      const a = search(excludeAttributes)
+      return !(a.indexOf(getNodeInputType(attr)) > -1)
+    })
```

Comment:
We iterate over the array three times where once would suffice. Normally, there is a trade-off with readability when fusing loop manipulations, but since in this case it's just three `filter`s, you can trivially combine them into one, no?